crates/typst/src/layout/flow.rs

```diff
@@ -1214,6 +1259,90 @@ impl<'a, 'e> FlowLayouter<'a, 'e> {
+        //
+        // We do this after placing all frames since they might not necessarily
+        // be ordered by height (e.g. you can have a `place(bottom)` followed
+        // by a paragraph, but the paragraph appears at the top), so we buffer
+        // all line numbers to later sort and deduplicate them based on how
+        // close they are to each other.
+        lines.sort_by_key(|line| line.y);
+
+        // TODO: Consider real line height
+        // Actually, we should store the height of each line next to the tag
+        // Then we deduplicate based on lines being inside others
+        const LINE_DISTANCE_THRESHOLD: Abs = Abs::raw(1.0);
+
+        // Buffer line number frames so we can align them horizontally later
+        // before placing, based on the width of the largest line number.
+        let mut line_numbers = vec![];
+        // Used for horizontal alignment.
+        let mut max_number_width = Abs::zero();
+        let mut prev_y = None;
+        for line in lines {
+            if prev_y
+                .is_some_and(|prev_y| (line.y - prev_y).abs() < LINE_DISTANCE_THRESHOLD)
+            {
+                // Lines are too close together. Display as the same line number.
+                continue;
+            }
+
+            // Note that line.y > prev_y due to sorting.
+            // Therefore, the check above ensures no lines too close together
+            // will cause too many different line numbers to appear.
+            prev_y = Some(line.y);
+            let current_column = self.finished.len() % self.columns;
+            let number_margin = if self.columns >= 2 && current_column + 1 == self.columns
+            {
+                // The last column will always place line numbers at the end
+                // margin. This should become configurable in the future.
+                OuterHAlignment::End.resolve(*self.styles)
+            } else {
+                line.marker.number_margin().resolve(*self.styles)
+            };
+
+            let number_align = line
+                .marker
+                .number_align()
+                .map(|align| align.resolve(*self.styles))
+                .unwrap_or_else(|| number_margin.inv());
+
+            let number_clearance = line.marker.number_clearance().resolve(*self.styles);
+            let number = self.layout_line_number(line.marker)?;
+            let number_x = match number_margin {
+                FixedAlignment::Start => -number_clearance,
+                FixedAlignment::End => size.x + number_clearance,
+                // Shouldn't be specifiable by the user due to 'OuterHAlignment'.
+                FixedAlignment::Center => unreachable!(),
+            };
+            let number_pos = Point::new(number_x, line.y);
+            max_number_width.set_max(number.width());
+            line_numbers.push((number_pos, number, number_align, number_margin));
+        }
+
+        for (mut pos, number, align, margin) in line_numbers {
```

Comment:
Any particular reason to first collect into `line_numbers` and then immediately iterate? Could we do it directly?

Reply:
This is due to collecting `max_number_width` which is needed for alignment.

An artifact of this, as can be seen in the `line-numbers-start-alignment` test, is that line numbers with `start` alignment in the `start` margin are probably aligned between themselves in the same page, but not between different pages. I'm not sure of the best way to fix this, or even if it would be worth it to do so, frankly (I expect `start` alignment on a `start` margin, or `end` on an `end` margin, to be a rare style choice anyway).

Reply:
makes sense